consumer/consumererror/error.go

```diff
@@ -0,0 +1,145 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package consumererror // import "go.opentelemetry.io/collector/consumer/consumererror"
+
+import (
+	"go.opentelemetry.io/collector/consumer/consumererror/internal/statusconversion"
+	"google.golang.org/grpc/status"
+)
+
+type Error struct {
+	errors []ErrorData
+}
+
+type ErrorData interface {
+	Error() string
+
+	Unwrap() error
+
+	// Returns nil if a status is not available.
+	HTTPStatus() (int, bool)
+
+	// Returns nil if a status is not available.
+	GRPCStatus() (*status.Status, bool)
+
+	Metadata() []any
+}
+
+type errorData struct {
+	error
+	httpStatus *int
+	grpcStatus *status.Status
+	metadata   []any
+}
+
+// ErrorOption allows annotating an Error with metadata.
+type ErrorOption func(error *errorData)
+
+// NewConsumerError wraps an error that happened while consuming telemetry
+// and adds metadata onto it to be passed back up the pipeline.
+func NewConsumerError(origErr error, options ...ErrorOption) error {
+	err := &Error{}
+	errData := &errorData{error: origErr}
+
+	for _, option := range options {
+		option(errData)
+	}
+
+	return err
+}
+
+var _ error = &Error{}
+
+func (e *Error) Error() string {
+	return e.errors[len(e.errors)-1].Error()
+}
+
+// Unwrap returns the wrapped error for use by `errors.Is` and `errors.As`.
+func (e *Error) Unwrap() []error {
+	errors := make([]error, 0, len(e.errors))
+
+	for _, err := range e.errors {
+		errors = append(errors, err)
+	}
+
+	return errors
+}
+
+func (e *Error) Combine(errs ...error) {
+	for _, err := range errs {
+		if ne, ok := err.(*Error); ok {
+			e.errors = append(e.errors, ne.errors...)
+		} else {
+			e.errors = append(e.errors, &errorData{error: err})
+		}
+	}
+}
+
+func (e *Error) Data() []ErrorData {
```

Comment:
I think we can potentially (not nesessarily now) add helper methods to `Error` that would encapsulate the complexity and pre-define the precedence of all the submitted options. For example, the following would always give statuses based on the information provided by the reporter
```
HTTPStatus() int
GRPCStatus() status.Status
```

Reply:
I'm okay with that. I would like to keep the core fairly unopinionated and have a separate wrapper to handle defaults, but if we think the defaults are fairly universal, I'm okay going that route.

Reply:
> I'm okay with that. I would like to keep the core fairly unopinionated and have a separate wrapper to handle defaults, but if we think the defaults are fairly universal, I'm okay going that route.

I'm afraid that it'll be to open for interpretation and we'll end up with different behaviors in the receivers. We still can expose the Data anyway

Reply:
My bad, reading it again, I think I misunderstood you. I would be fine adding methods to `Error` specifically that aggregates information across all the different `ErrorData` objects. I would like to do that in a follow-up if possible, though.

Reply:
Sounds good 👍